### ISSUE TITLE
Views with non-zero begin index

### ISSUE THREAD

**Issue by tcfisher** [kokkos/kokkos#567]
For structured meshes, we have some use cases where we we'd like to match indices on a volume block with a view of the same rank, but only on the surface.

For example, if we have a view representing a volume block, a(7,5,6), we'd like to be able to create a view of the same rank representing a surface of the volume block, b(1,5,6), but with matching indices for the last index, b(6:6,0:4,0:5). We have a workaround for this using a wrapper with support for the use cases we care about so far, but it would be nice if there was some native Kokkos support for this use case.

**Comment by crtrott** [comment]
What do you hope to gain from that, i.e. why don't pass the original view on?

For me it looks like you want to do this:

```C++
void foo_surface(View<double***> surface, int k) {
  for(i=0;i<N;i++)
  for(j=0;j<M;j++)
  surface(k,i,j) = ...
}

void foo(View<double***> volume) {
  {
     auto surface = subview_offset(volume,make_pair(0,0),ALL,ALL);
     foo_surface(surface,0);
  }
  {
     auto surface = subview_offset(volume,make_pair(6,6),ALL,ALL);
     foo_surface(surface,6);
  }
}
```

If we could make the subview thing have no overhead it would be literally the same as:
```C++
void foo_surface(View<double***> surface, int k) {
  for(i=0;i<N;i++)
  for(j=0;j<M;j++)
  surface(k,i,j) = ...
}

void foo(View<double***> volume) {
  {
     foo_surface(volume,0);
  }
  {
     foo_surface(volume,6);
  }
}
```

Indeed the second variant would be cheaper, because I don't use registers for the subview.

**Comment by tcfisher** [comment]
The surface may be completely separate data from the data in the volume block, so subview is not applicable. E.g. b is only defined on a surface. For example, skin friction is only defined on a wall, whereas the gradients we use to compute it are defined everywhere in the volume. The goal is be able to index into the surface arrays using the same index space as the volume array without having to take on memory overhead of an array for the whole volume. It would be inappropriate to allocate a full volume field in this case for something that is only defined on a surface, which is what would be needed to subview, right?

**Comment by crtrott** [comment]
Btw. I thought about the non-zero begin index before. Mostly actually to have things go from -B to N+B for boundaries i.e. my owned volume is 0-N and the boundaries are around it. But there is some overhead involved in this. Basically I need to store the offsets somewhere, because I can't simply move my pointer (i.e. I can't just access ptr[-512]). This would break our reference counting, creating subviews, and bounds checking. That means the size of the views is now basically the same as that of Strided views because I need to store length and offset for each dimension.

**Comment by crtrott** [comment]
Ok usecase was clarified in phone call.

**Comment by crtrott** [comment]
We want LayoutLeftOffset, LayoutRightOffset, and LayoutStridedOffset. View constructions would work via providing a Layout instance. The Layout constructors take additional offset arguments for each dimension.

**Comment by crtrott** [comment]
I don't think we can finish this this CY - at least not with all the machinery of subviews and interoperability with subviews.

**Comment by crtrott** [comment]
Subview question: should a subview start at 0 or should it maintain the original views offsets or do you want both.

```C++
int B0 = 2;
int B1 = 3;
int N0 = 2*B0 + N;
int N1 = 2*B1 + N;

LayoutLeftOffset layout({N0,-B0},{N1,-B1});
View<double**,LayoutLeftOffset> a("A",layout);

View<double**,LayoutLeftOffset> a_(a,make_pair(1,N-1),ALL);

// How do you index a_inner_1?
// a(1,...) ... a(N-2, ...) 
// or
// a(0,...) ... a(N-3, ...)

**Comment by tcfisher** [comment]
If possible I'd use Fortran90 to guide us. In which case a subview would take the subview offsets for the subset dimension and the original view offsets for the other dimensions. E.g. By default I'd prefer that the inner layout of a_ would go from 1 to N-1. I could see how this would be potentially problematic if you didn't know a view had an offset layout. The general case would make it optionally take it's own layout to remap to any other set with the same size?

**Comment by hcedwar** [comment]
The wrapper strategy is less invasive and likely to be safer.  Consider hosting such a wrapper within Kokkos itself.  Can draw upon UIntah experience as well.

**Comment by crtrott** [comment]
I am not sure I agree. There are performance considerations to not have it directly implemented, and in particular getting the subview functionality right is tough work for somebody not intimately familiar with the internals of Kokkos Views and ViewMapping. I am very much in favor of having this being implemented as a Layout.

**Comment by hcedwar** [comment]
The feature request is a fundamental change to our [0..N_0)x...[0..N_r) index space abstraction.  There are to-be-determined (numerous) points in the View infrastructure that have this assumption.  For example, subview, bounds checking, deep copy of non-conforming layouts, ...
A systematic re-design to allow a Layout to define its index space domain is required.  This abstraction change is not a trivial endeavor.

**Comment by ramanan** [comment]
I am interested in a non-zero begin index as well. My use case is what @crtrott mentioned above. The index needs to go from -B to N-1+B. This will make it easy to port legacy code and make the code readable. 

I see that a wrapper has been mentioned in this thread. Is there an example wrapper that I can build upon?

**Comment by stanmoore1** [comment]
I could also have used this previously as well in LAMMPS--instead I had to add the offset to all of the indices so they started at zero which was a pain.

**Comment by swbova** [comment]
@crtrott @ndellingwood and I had a meeting today to discuss the design of this enhancement.  Originally, we thought that this would be implemented via a new Layout.  Upon reflection, this approach has at least two significant issues:

1. Existing Layouts have zero start index and extents.  adding a "begin" and an extent is a nontrivial change, especially with respect to kokkos-kernels.  Eg dot products would have to be modified to go from begin to end, and interoperability with Views with standard Layouts would be a chore.
2. C++ standard uses zero start plus extents for containers, and making the existing View a departure from that is not desirable.

Hence, we decided to implement a new View called OffsetView.  This new view will be in containers, and

-  can use the existing map and track classes for its member data 
- will have new member data to hold the offsets, which are defined as starting indices that can be positive or negative
-  can easily be constructed from a view and offsets
- can be constructed using offsets and a Layout
-  can be easily converted to a View
- constructors
  -  OffsetView(name, (imin, imax), (jmin,jmax)... )
  - OffsetView(name, Layout, imin, jmin, kmin)
    - LayoutLeft
    - LayoutRight
    - LayoutStrided
  - OffsetView(View, imin, jmin, kmin) 
- will require specializations of Subview creators 
  - if ALL is passed, then offsets are preserved for those indices
  - if a range is passed then offsets are dropped and access is zero-based
- Subview implementation will work by converting given OffsetView to a View, generate new Arg pack, call the standard Subview function, then convert resulting View to OffsetView

**Comment by tcfisher** [comment]
@swbova That all sounds fine to me. The subview approach seems to be sound and it covers clearly the use cases I have in mind. Thanks.

**Comment by ibaned** [comment]
I think this has been implemented by pull request #1623
